src/validation.cpp

```diff
@@ -3686,6 +3691,124 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+{
+    // TODO: merge with ConnectBlock
+    CBlock block;
+    if (!ReadBlockFromDisk(block, pindex, params.GetConsensus())) {
+        return error("ReplayBlock(): ReadBlockFromDisk failed at %d, hash=%s", pindex->nHeight, pindex->GetBlockHash().ToString());
+    }
+
+    for (const CTransactionRef& tx : block.vtx) {
+        if (!tx->IsCoinBase()) {
+            for (const CTxIn &txin : tx->vin) {
+                inputs.ModifyCoins(txin.prevout.hash)->Spend(txin.prevout.n);
+            }
+        }
+        // We cannot use ModifyNewCoins here, as the entry may exist in the cache already.
+        inputs.ModifyCoins(tx->GetHash())->FromTx(*tx, pindex->nHeight);
+    }
+    return true;
+}
+
+/** Unapply the effects of a block on the utxo cache, ignoring that it may already have been applied. */
+static bool RollbackBlock(const CBlockIndex* pindex, CCoinsViewCache& inputs, const CChainParams& params)
+{
+    // TODO: Merge with DisconnectBlock
+    CBlock block;
+    if (!ReadBlockFromDisk(block, pindex, params.GetConsensus())) {
+        return error("RollbackBlock(): ReadBlockFromDisk failed at %d, hash=%s", pindex->nHeight, pindex->GetBlockHash().ToString());
+    }
+
+    CBlockUndo blockUndo;
+    CDiskBlockPos pos = pindex->GetUndoPos();
+    if (pos.IsNull())
+        return error("RollbackBlock(): no undo data available");
+    if (!UndoReadFromDisk(blockUndo, pos, pindex->pprev->GetBlockHash()))
+        return error("RollbackBlock(): failure reading undo data");
+
+    if (blockUndo.vtxundo.size() + 1 != block.vtx.size())
+        return error("RollbackBlock(): block and undo data inconsistent");
+
+    for (int i = block.vtx.size() - 1; i >= 0; --i) {
+        const CTransaction& tx = *block.vtx[i];
+        inputs.ModifyCoins(tx.GetHash())->Clear();
+        if (i != 0) {
+            CTxUndo &txundo = blockUndo.vtxundo[i - 1];
+            if (txundo.vprevout.size() != tx.vin.size()) {
+                return error("RollbackBlock(): block and undo data inconsistent");
+            }
+            for (size_t n = 0; n < txundo.vprevout.size(); ++n) {
+                ApplyTxInUndo(txundo.vprevout[n], inputs, COutPoint(tx.GetHash(), n)); // Ignore return value
+            }
+        }
+    }
+    return true;
+}
+
+
+bool ReplayBlocks(const CChainParams& params, CCoinsView* view)
+{
+    LOCK(cs_main);
+
+    CCoinsViewCache cache(view);
+
+    uint256 hashBest = view->GetBestBlock();
+    if (!hashBest.IsNull()) return true;
+    std::vector<uint256> hashHeads = view->GetHeadBlocks();
+    if (hashHeads.empty()) return true;
+
+    uiInterface.ShowProgress(_("Replaying blocks..."), 0);
+    LogPrintf("Replaying blocks\n");
+
+    std::vector<const CBlockIndex*> pindexHeads;
+    const CBlockIndex* pindexFork = nullptr;
+    bool fGenesis = false;
+
+    // Find last common ancestor of all heads.
+    for (const uint256& hash : hashHeads) {
+        if (hash.IsNull()) {
```

Comment:
I am confused about all this `fGenesis` variable. I guess this is related to my other comment on https://github.com/bitcoin/bitcoin/pull/10148/commits/02dfa4a54da44e40b1d29687cc68885f3a306ef3#r111875593

I would not expect, pindexFork to be null, as the work fork point should be the genesis block.